### PR TITLE
Add append_slice for GenericStringBuilder

### LINKED ISSUE #352: Shortcut for utf8 bytes append
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I want to append utf8 bytes from a trusted source to a `GenericStringBuilder` directly.

**Describe the solution you'd like**
I'd add a `fn append_slice(&mut self, value: &[u8])` helper function to the GenericStringBuilder struct.

**Describe alternatives you've considered**
Currently this is a workaround:
`use arrow::array::GenericStringBuilder.append_value(std::str::from_utf8_unchecked(my_bytes))`

Related links:
- https://docs.rs/arrow/4.1.0/src/arrow/array/builder.rs.html#1189-1196
- https://doc.rust-lang.org/src/core/str/converts.rs.html#157-165

### PR DESCRIPTION
**Pull request by alippai**
# Which issue does this PR close?

Closes #352.

 # Rationale for this change

Developer experience and perhaps performance. 
 

# What changes are included in this PR?

A new function `append_slice(&mut self, value: &[u8])` is added to `GenericStringBuilder`

# Are there any user-facing changes?
Only the new function, it's a non-breaking addition.
A test and simple docs are included.

### CODE REVIEW

**Comment by alippai** [comment]
Alternatively I can turn this into append_slice_unchecked and create an append_slice which adds `run_utf8_validation(...) -> Result<>` like seen [here](https://doc.rust-lang.org/src/core/str/converts.rs.html#86), but since this is a builder, I'd prefer keeping it this way and adding an utf8 validation kernel later.

**Comment by ritchie46** [comment]
This is `unsafe` without an `unsafe` keyword. Rust assumes valid utf8 for all `str` and `String` values, consequently arrow assumes this as well.

If you want to create a `&str` from a slice without checking validity, you should use [https://doc.rust-lang.org/std/str/fn.from_utf8_unchecked.html](https://doc.rust-lang.org/std/str/fn.from_utf8_unchecked.html) which requires unsafe code.

That `&str` can then be used in `GenericStringBuilder::append`. I don't think you should add an `unsafe` API for something that's already in the std library.

**Comment by alippai** [comment]
@ritchie46 thanks, it makes sense. I'll rework it a little bit.

**Comment by alippai** [comment]
@ritchie46 updated the API with correct name + unsafe, also added benchmark.
```
bench_string/bench_string                                                                            
                        time:   [10.194 ms 10.259 ms 10.333 ms]
                        thrpt:  [1.1341 GiB/s 1.1423 GiB/s 1.1496 GiB/s]
bench_string/bench_string_slice                                                                            
                        time:   [19.182 ms 19.303 ms 19.418 ms]
                        thrpt:  [617.98 MiB/s 621.68 MiB/s 625.59 MiB/s]
bench_string/bench_string_slice_unchecked                                                                            
                        time:   [7.2051 ms 7.2217 ms 7.2388 ms]
                        thrpt:  [1.6189 GiB/s 1.6227 GiB/s 1.6265 GiB/s]
```

File: arrow/src/array/builder.rs [hunk 1]
```diff
@@ -1246,6 +1247,33 @@ impl<OffsetSize: StringOffsetSizeTrait> GenericStringBuilder<OffsetSize> {
...
     }
 
+    /// Appends a UTF-8 slice into the builder.
+    ///
+    /// Automatically calls the `append` method to delimit the slice appended in as a
+    /// distinct array element.
+    #[inline]
+    pub fn append_slice(&mut self, value: &[u8]) -> Result<()> {
+        // Do not append invalid UTF-8
+        from_utf8(value).or(Err(ArrowError::InvalidArgumentError(
```
Comment (Dandandan) [inline arrow/src/array/builder.rs:1257]: I think this `or` should be using `or_else` here otherwise the error path is always evaluated

**Comment by alippai** [comment]
Checking with godbolt and criterion this is basically just a wrapper around `from_utf8` and `from_utf8_unchecked`. In light of this I'm closing this PR, maybe the docs can be extended if needed.

**Pull request closed without merging**